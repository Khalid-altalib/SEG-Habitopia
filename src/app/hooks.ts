import {
  useDispatch as reduxUseDispatch,
  useSelector as reduxUseSelector,
} from "react-redux";
import type { TypedUseSelectorHook } from "react-redux";
import type { RootState, AppDispatch } from "./store";

export const useDispatch: () => AppDispatch = reduxUseDispatch;
export const useSelector: TypedUseSelectorHook<RootState> = reduxUseSelector;
