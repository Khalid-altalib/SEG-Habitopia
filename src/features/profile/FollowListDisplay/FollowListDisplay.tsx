// React
import React from "react";

// React Native
import { TouchableOpacity } from "react-native";

// React Navigation
import { useNavigation } from "@react-navigation/native";
import { NativeStackNavigationProp } from "@react-navigation/native-stack";

// Expo Linear Gradient
import { LinearGradient } from "expo-linear-gradient";

// Native Base
import { Box, Center, theme } from "native-base";

// Habitopia
import { useDispatch, useSelector } from "@app/hooks";
import { RootParams, TextType } from "types";
import { fetchFollowList } from "../profileSlice";
import Text from "@components/Text";

type Props = {
  followListMode: string;
  followCount: number;
};

/**
 * Displays a box containing the follow list count and a label for a particular follow list mode.
 *
 * @returns The component representing the box.
 */
const FollowListDisplay = (props: Props) => {
  const { followListMode, followCount } = props;
  const profile = useSelector((state) => state.profile?.profile);

  const navigation = useNavigation<NativeStackNavigationProp<RootParams>>();

  const dispatch = useDispatch();

  const handlePress = () => {
    navigation.push("FollowList", { followListMode: followListMode });
    dispatch(fetchFollowList({ followListMode, profileID: profile?.userId }));
  };

  // Gets the capitalized form of the follow list type
  const followListLabel =
    followListMode[0].toUpperCase() + followListMode.slice(1);

  return (
    <Box overflow="hidden" rounded="lg" flexGrow={1}>
      <TouchableOpacity onPress={handlePress} testID="follow-list-display">
        <LinearGradient
          colors={[theme.colors.darkBlue[500], theme.colors.purple[700]]}
          start={[0, 0]}
          end={[1, 1]}
          style={{
            padding: 12.5,
          }}
        >
          <Center>
            <Text type={TextType.Subheading} color="white">
              {followCount}
            </Text>
            <Text type={TextType.Small} color="white">
              {followListLabel}
            </Text>
          </Center>
        </LinearGradient>
      </TouchableOpacity>
    </Box>
  );
};

export default FollowListDisplay;
