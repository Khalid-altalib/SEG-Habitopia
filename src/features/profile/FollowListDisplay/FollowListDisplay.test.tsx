import { useDispatch, useSelector } from "@app/hooks";
import { useNavigation } from "@react-navigation/native";
import {render, fireEvent } from '@testing-library/react-native';
import { Box, Center, theme } from "native-base";
import React from "react";
import TestingWrapperNavigation from '@app/testingWrapperWithNavigation';
import configureStore from "redux-mock-store";
import FollowListDisplay from "./FollowListDisplay";
import thunk from "redux-thunk";

const mockedDispatch = jest.fn();
const mockedUseSelector = useSelector as jest.Mock;
const mockedUseNavigation = useNavigation as jest.Mock;

mockedUseSelector.mockReturnValue({
  profile: {
    userId: "123",
  },
});

mockedUseNavigation.mockReturnValue({
  push: jest.fn(),
});

jest.mock('../../../app/hooks', () => ({
    useSelector: jest.fn(),
    useDispatch: () => mockedDispatch,
}));

jest.mock('@react-navigation/native', () => ({
    ...jest.requireActual('@react-navigation/native'),
    useNavigation: jest.fn(),
}));


describe('FollowListDisplay', () => {

    const mockState = {
        profile: {
          profile: {
            userId: '123',
            name: 'Test User',
            followerCount: 10,
            followingCount: 20,
            biography: 'Test Biography',
          },
        },
      };
    
      const mockStore = configureStore([thunk])(mockState);
      let wrapper: any;
      beforeEach(() => {
        (useSelector as jest.Mock).mockImplementation((selector) => selector(mockStore.getState()));
        wrapper = render(
          <TestingWrapperNavigation store={mockStore}>
            <FollowListDisplay followListMode={"follower"} followCount={10} />
          </TestingWrapperNavigation>
        );
      });
      it("renders correctly", () => {
        expect(wrapper).toMatchSnapshot();
      });

      it("should push navigation to followList when handlePress is called", () => {
        const followListButton = wrapper.getByTestId('follow-list-display');
        fireEvent.press(followListButton);
        expect(mockedUseNavigation().push).toHaveBeenCalledWith("FollowList", {followListMode: "follower"});
        expect(mockedDispatch).toHaveBeenCalled();
        expect(wrapper).toMatchSnapshot();
      });
});