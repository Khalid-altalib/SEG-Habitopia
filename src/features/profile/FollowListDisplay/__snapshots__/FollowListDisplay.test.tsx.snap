// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`FollowListDisplay renders correctly 1`] = `
<RNCSafeAreaProvider
  onInsetsChange={[Function]}
  style={
    [
      {
        "flex": 1,
      },
      undefined,
    ]
  }
>
  <View
    style={
      [
        {
          "flex": 1,
        },
        undefined,
      ]
    }
  >
    <RNSScreenStack
      style={
        {
          "flex": 1,
        }
      }
    >
      <RNSScreen
        collapsable={false}
        forwardedRef={[Function]}
        gestureResponseDistance={
          {
            "bottom": -1,
            "end": -1,
            "start": -1,
            "top": -1,
          }
        }
        nativeBackButtonDismissalEnabled={false}
        onAppear={[Function]}
        onDisappear={[Function]}
        onDismissed={[Function]}
        onHeaderBackButtonClicked={[Function]}
        onNativeDismissCancelled={[Function]}
        onTransitionProgress={[Function]}
        onWillDisappear={[Function]}
        replaceAnimation="push"
        stackPresentation="push"
        style={
          {
            "bottom": 0,
            "left": 0,
            "position": "absolute",
            "right": 0,
            "top": 0,
          }
        }
        swipeDirection="horizontal"
      >
        <RNSScreenStackHeaderConfig
          backButtonInCustomView={false}
          backgroundColor="rgb(255, 255, 255)"
          color="rgb(0, 122, 255)"
          direction="ltr"
          disableBackButtonMenu={false}
          hidden={false}
          hideBackButton={false}
          largeTitleHideShadow={false}
          title="TestScreen"
          titleColor="rgb(28, 28, 30)"
          topInsetEnabled={false}
          translucent={false}
        />
        <View
          accessibilityElementsHidden={false}
          importantForAccessibility="auto"
          style={
            {
              "flex": 1,
              "flexDirection": "column-reverse",
            }
          }
        >
          <View
            style={
              [
                {
                  "flex": 1,
                },
                {
                  "backgroundColor": "rgb(242, 242, 242)",
                },
                undefined,
              ]
            }
          >
            <View
              dataSet={{}}
              style={
                {
                  "borderRadius": 8,
                  "flexGrow": 1,
                  "overflow": "hidden",
                }
              }
            >
              <View
                accessible={true}
                collapsable={false}
                focusable={true}
                onClick={[Function]}
                onResponderGrant={[Function]}
                onResponderMove={[Function]}
                onResponderRelease={[Function]}
                onResponderTerminate={[Function]}
                onResponderTerminationRequest={[Function]}
                onStartShouldSetResponder={[Function]}
                style={
                  {
                    "opacity": 1,
                  }
                }
                testID="follow-list-display"
              >
                <ViewManagerAdapter_ExpoLinearGradient
                  proxiedProperties={
                    {
                      "colors": [
                        4278220774,
                        4286456526,
                      ],
                      "endPoint": [
                        1,
                        1,
                      ],
                      "locations": undefined,
                      "startPoint": [
                        0,
                        0,
                      ],
                    }
                  }
                  style={
                    {
                      "padding": 12.5,
                    }
                  }
                >
                  <View
                    dataSet={{}}
                    style={
                      {
                        "alignItems": "center",
                        "display": "flex",
                        "justifyContent": "center",
                      }
                    }
                  >
                    <Text
                      style={
                        [
                          {
                            "color": "white",
                          },
                          {
                            "fontFamily": undefined,
                            "fontSize": 23,
                          },
                          undefined,
                        ]
                      }
                    >
                      10
                    </Text>
                    <Text
                      style={
                        [
                          {
                            "color": "white",
                          },
                          {
                            "fontFamily": undefined,
                            "fontSize": 15,
                          },
                          undefined,
                        ]
                      }
                    >
                      Follower
                    </Text>
                  </View>
                </ViewManagerAdapter_ExpoLinearGradient>
              </View>
            </View>
          </View>
        </View>
      </RNSScreen>
    </RNSScreenStack>
  </View>
</RNCSafeAreaProvider>
`;

exports[`FollowListDisplay should push navigation to followList when handlePress is called 1`] = `
<RNCSafeAreaProvider
  onInsetsChange={[Function]}
  style={
    [
      {
        "flex": 1,
      },
      undefined,
    ]
  }
>
  <View
    style={
      [
        {
          "flex": 1,
        },
        undefined,
      ]
    }
  >
    <RNSScreenStack
      style={
        {
          "flex": 1,
        }
      }
    >
      <RNSScreen
        collapsable={false}
        forwardedRef={[Function]}
        gestureResponseDistance={
          {
            "bottom": -1,
            "end": -1,
            "start": -1,
            "top": -1,
          }
        }
        nativeBackButtonDismissalEnabled={false}
        onAppear={[Function]}
        onDisappear={[Function]}
        onDismissed={[Function]}
        onHeaderBackButtonClicked={[Function]}
        onNativeDismissCancelled={[Function]}
        onTransitionProgress={[Function]}
        onWillDisappear={[Function]}
        replaceAnimation="push"
        stackPresentation="push"
        style={
          {
            "bottom": 0,
            "left": 0,
            "position": "absolute",
            "right": 0,
            "top": 0,
          }
        }
        swipeDirection="horizontal"
      >
        <RNSScreenStackHeaderConfig
          backButtonInCustomView={false}
          backgroundColor="rgb(255, 255, 255)"
          color="rgb(0, 122, 255)"
          direction="ltr"
          disableBackButtonMenu={false}
          hidden={false}
          hideBackButton={false}
          largeTitleHideShadow={false}
          title="TestScreen"
          titleColor="rgb(28, 28, 30)"
          topInsetEnabled={false}
          translucent={false}
        />
        <View
          accessibilityElementsHidden={false}
          importantForAccessibility="auto"
          style={
            {
              "flex": 1,
              "flexDirection": "column-reverse",
            }
          }
        >
          <View
            style={
              [
                {
                  "flex": 1,
                },
                {
                  "backgroundColor": "rgb(242, 242, 242)",
                },
                undefined,
              ]
            }
          >
            <View
              dataSet={{}}
              style={
                {
                  "borderRadius": 8,
                  "flexGrow": 1,
                  "overflow": "hidden",
                }
              }
            >
              <View
                accessible={true}
                collapsable={false}
                focusable={true}
                onClick={[Function]}
                onResponderGrant={[Function]}
                onResponderMove={[Function]}
                onResponderRelease={[Function]}
                onResponderTerminate={[Function]}
                onResponderTerminationRequest={[Function]}
                onStartShouldSetResponder={[Function]}
                style={
                  {
                    "opacity": 1,
                  }
                }
                testID="follow-list-display"
              >
                <ViewManagerAdapter_ExpoLinearGradient
                  proxiedProperties={
                    {
                      "colors": [
                        4278220774,
                        4286456526,
                      ],
                      "endPoint": [
                        1,
                        1,
                      ],
                      "locations": undefined,
                      "startPoint": [
                        0,
                        0,
                      ],
                    }
                  }
                  style={
                    {
                      "padding": 12.5,
                    }
                  }
                >
                  <View
                    dataSet={{}}
                    style={
                      {
                        "alignItems": "center",
                        "display": "flex",
                        "justifyContent": "center",
                      }
                    }
                  >
                    <Text
                      style={
                        [
                          {
                            "color": "white",
                          },
                          {
                            "fontFamily": undefined,
                            "fontSize": 23,
                          },
                          undefined,
                        ]
                      }
                    >
                      10
                    </Text>
                    <Text
                      style={
                        [
                          {
                            "color": "white",
                          },
                          {
                            "fontFamily": undefined,
                            "fontSize": 15,
                          },
                          undefined,
                        ]
                      }
                    >
                      Follower
                    </Text>
                  </View>
                </ViewManagerAdapter_ExpoLinearGradient>
              </View>
            </View>
          </View>
        </View>
      </RNSScreen>
    </RNSScreenStack>
  </View>
</RNCSafeAreaProvider>
`;
